### ISSUE TITLE
Tesseract Training Guide

### ISSUE THREAD

**Issue by zhao-lun** [Shreeshrii/tess5train-fonts#4]
Hi Shree,

Forgive me, as im new to LSTM and Tesseract. 
I ran tesseract API on windows, It works best with default tesseract_best/eng.traineddata (sometime it perform better on your trained weights(/tessdata_shreetest/engrestrict_best.traineddata)
However, the result yield is not very satisfactory. (about 55% decoded correctly) (image failed due to low contrast and unknown reasons)
Therefore, I seeks to fine-tune it with some training. Then i realizes the intended training datasets aren't using images, rather it generates image from text.(not very sure as the documentation are extremely confusing).
Q1: Can i train it with my images(contrasts varries), in hope it can perform better on my datasets?

Example:
[aaa_1_2_0](https://user-images.githubusercontent.com/32569157/59243065-bf7dd080-8c40-11e9-905e-c17f75be5123.jpg)
Default weights: ASA
Ground Truth: A8A

**Comment by Shreeshrii** [comment]
Better results are sometimes received by pre-processing the images rather than training. See improving quality page in wiki.

Regarding training with images, see ocrd-train project. You can also use wordstrbox option to create box/tiff pairs using images.

**Comment by zhao-lun** [comment]
OKay, Thanks ! I'll try that.
Note: I tried enhance the contrast on the image, Otsu binarization, median filter, etc, etc. None of them works so far.

**Comment by Shreeshrii** [comment]
https://github.com/tesseract-ocr/tesseract/wiki/ImproveQuality

https://github.com/OCR-D/ocrd-train

https://github.com/tesseract-ocr/tesseract/issues/2357#issuecomment-477261248

**Comment by zhao-lun** [comment]
Thanks @Shreeshrii , I have tried them(in the wiki/ImproveQuality)
I wanted to ask, the bounding box provided by Tesseract is not accurate? it's a bug or it's normal?
Tried OCRD-train(100images):
Result aren't that good, maybe because small dataset

Sample:
Ori: 
![aaaaaa](https://user-images.githubusercontent.com/32569157/59328226-6f753b80-8d1e-11e9-9143-eea49ec54d2f.jpg)


Detection:
![aaaa](https://user-images.githubusercontent.com/32569157/59328213-684e2d80-8d1e-11e9-8ad0-3ada45b8b1f9.jpg)
Note: I did some Otsu Thres there from the original image
for this image, Tesseract predicts correctly, just the bounding boxes are incorrect

**Comment by Shreeshrii** [comment]
If accuracy of bounding boxes is important then you should use 3.0x tesseract with tessdata. The 4.0x version is trained on line images and accuracy of bounding box positions is not important.

Or you could use a different method eg. opencv etc to get the bounding boxes.

**Comment by zthdsb** [comment]
Excause me, Is there any requirement for training data?Do you have to use picture that contain multiple characters?Is it ok to use picture that contain only one word?

**Comment by Shreeshrii** [comment]
@johnnydept The latest code from master branch has a better algorithm for bounding boxes.  Please give it a try.

```
tesseract MC34704.jpg - --oem 1 --psm 6 --tessdata-dir ~/tessdata_best  -c hocr_char_boxes=1 hocr


<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
    "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
 <head>
  <title></title>
  <meta http-equiv="Content-Type" content="text/html;charset=utf-8"/>
  <meta name='ocr-system' content='tesseract 5.0.0-alpha-322-g74ac' />
  <meta name='ocr-capabilities' content='ocr_page ocr_carea ocr_par ocr_line ocrx_word ocrp_wconf'/>
 </head>
 <body>
  <div class='ocr_page' id='page_1' title='image "MC34704.jpg"; bbox 0 0 235 55; ppageno 0'>
   <div class='ocr_carea' id='block_1_1' title="bbox 10 9 216 45">
    <p class='ocr_par' id='par_1_1' lang='eng' title="bbox 10 9 216 45">
     <span class='ocr_line' id='line_1_1' title="bbox 10 9 216 45; baseline 0.005 -1; x_size 46.833332; x_descenders 11.708333; x_ascenders 11.708333">
      <span class='ocrx_word' id='word_1_1' title='bbox 10 9 216 45; x_wconf 63'>
       <span class='ocrx_cinfo' title='x_bboxes 10 10 42 44; x_conf 98.883591'>M</span>
       <span class='ocrx_cinfo' title='x_bboxes 46 9 72 44; x_conf 98.706924'>C</span>
       <span class='ocrx_cinfo' title='x_bboxes 77 9 100 44; x_conf 98.782837'>3</span>
       <span class='ocrx_cinfo' title='x_bboxes 105 10 128 45; x_conf 99.027031'>4</span>
       <span class='ocrx_cinfo' title='x_bboxes 134 9 159 43; x_conf 98.472176'>7</span>
       <span class='ocrx_cinfo' title='x_bboxes 165 9 188 44; x_conf 94.743233'>6</span>
       <span class='ocrx_cinfo' title='x_bboxes 192 10 216 45; x_conf 98.931549'>4</span>
      </span>
     </span>
    </p>
   </div>
  </div>
 </body>
</html>
```

**Comment by Shreeshrii** [comment]
>Do you have to use picture that contain multiple characters?Is it ok to use picture that contain only one word?

@zthdsb  The neural net LSTM engine is supposed to work on line images. The recommnded method for training is to use text and fonts with the text2image program to generate the box/tiff pairs.

It is possible to use images. Lots of people use https://github.com/OCR-D/ocrd-train for training with images. You can ask for more info on tesseract-ocr forum.

**Comment by zhao-lun** [comment]
Hi @Shreeshrii ,
So glad that the recent version fix the bounding boxes. 
Well, i found out that the recent version supports user pattern file. 
However, i faced some problem
1) The files, User_pattern_file. only load during init?
Reason is, i have bundle of images, previous how i do it is, i only api->init once e.g
`if (api->Init(NULL, "eng", tesseract::OEM_LSTM_ONLY)) 
{
}`
it ran all images then it frees the api->init. 
Problem:
Is it necessary to init every time just to read the config file?? 
Otherwise, i have to change my code structure when it keep initialize for new image for new config file.

**Comment by Shreeshrii** [comment]
I do not know much about the api. I suggest you post in tesseract-ocr google group so that others can respond.

**Comment by zhao-lun** [comment]
thanks @Shreeshrii , will do !
